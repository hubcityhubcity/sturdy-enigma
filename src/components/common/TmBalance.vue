<template>
  <div class="balance-header">
    <div class="values-container">
      <div class="total-atoms">
        <h3>Total {{ stakingDenom }}</h3>
        <h2 class="total-atoms__value">
          {{ overview.totalStake | shortDecimals | noBlanks }}
        </h2>
      </div>

      <div class="row small-container">
        <div class="available-atoms">
          <h3>Available {{ stakingDenom }}</h3>
          <h2>{{ overview.liquidStake | shortDecimals | noBlanks }}</h2>
        </div>

        <div v-if="overview.totalRewards" class="rewards">
          <h3>Total Rewards</h3>
          <h2>{{ overview.totalRewards | shortDecimals | noBlanks }}</h2>
        </div>
      </div>
    </div>
    <div class="button-container">
      <TmBtn
        class="send-button"
        value="Send"
        type="secondary"
        @click.native="onSend()"
      />
      <TmBtn
        id="withdraw-btn"
        :disabled="!readyToWithdraw"
        class="withdraw-rewards"
        value="Claim Rewards"
        @click.native="readyToWithdraw && onWithdrawal()"
      />
    </div>

    <SendModal ref="SendModal" :denom="stakingDenom" />
    <ModalWithdrawRewards ref="ModalWithdrawRewards" />
  </div>
</template>
<script>
import { shortDecimals } from "scripts/num"
import refetchNetworkOnly from "scripts/refetch-network-only"
import { noBlanks } from "src/filters"
import TmBtn from "common/TmBtn"
import SendModal from "src/ActionModal/components/SendModal"
import ModalWithdrawRewards from "src/ActionModal/components/ModalWithdrawRewards"
import { UserTransactionAdded } from "src/gql"
import { mapGetters } from "vuex"
import gql from "graphql-tag"
export default {
  name: `tm-balance`,
  components: {
    TmBtn,
    SendModal,
    ModalWithdrawRewards
  },
  filters: {
    shortDecimals,
    noBlanks
  },
  data() {
    return {
      overview: {},
      stakingDenom: ""
    }
  },
  computed: {
    ...mapGetters([`address`, `network`]),
    // only be ready to withdraw of the validator rewards are loaded and the user has rewards to withdraw
    // the validator rewards are needed to filter the top 5 validators to withdraw from
    readyToWithdraw() {
      return this.overview.totalRewards > 0
    }
  },
  methods: {
    onWithdrawal() {
      this.$refs.ModalWithdrawRewards.open()
    },
    onSend() {
      this.$refs.SendModal.open()
    }
  },
  apollo: {
    overview: {
      query: gql`
        query overview($networkId: String!, $address: String!) {
          overview(networkId: $networkId, address: $address) {
            totalRewards
            liquidStake
            totalStake
          }
        }
      `,
      variables() {
        /* istanbul ignore next */
        return {
          networkId: this.network,
          address: this.address
        }
      },
      update(data) {
        /* istanbul ignore next */
        return {
          ...data.overview,
          totalRewards: Number(data.overview.totalRewards)
        }
      },
      skip() {
        return !this.address
      }
    },
    stakingDenom: {
      query: gql`
        query Network($networkId: String!) {
          network(id: $networkId) {
            id
            stakingDenom
          }
        }
      `,
      variables() {
        return {
          networkId: this.network
        }
      },
      update(data) {
        /* istanbul ignore next */
        return data.network.stakingDenom
      }
    },
    $subscribe: {
      userTransactionAdded: {
        variables() {
          return {
            networkId: this.network,
            address: this.address
          }
        },
        skip() {
          return !this.address
        },
        query: UserTransactionAdded,
        result() {
          // query if successful or not as even an unsuccessful tx costs fees
          refetchNetworkOnly(this.$apollo.query.overview)
        }
      }
    }
  }
}
</script>
<style scoped>
.balance-header {
  display: flex;
  flex-direction: column;
  width: 100%;
}

.values-container {
  display: flex;
  position: relative;
  width: 100%;
  padding: 1rem 2rem;
  flex-direction: column;
}

.values-container h2 {
  font-size: 24px;
  font-weight: 500;
  line-height: 24px;
  color: var(--bright);
}

.values-container h3 {
  font-size: var(--sm);
  font-weight: 400;
  white-space: nowrap;
}

.total-atoms,
.available-atoms,
.rewards {
  padding-right: 2.5rem;
}

.rewards h2 {
  color: var(--success);
  font-size: var(--m);
}

.available-atoms h2 {
  font-size: var(--m);
  line-height: 20px;
}

.button-container {
  display: flex;
  align-items: center;
  padding: 0.5rem 2rem;
  width: 100%;
  border-bottom: 1px solid var(--bc-dim);
  border-top: 1px solid var(--bc-dim);
  margin-bottom: 2rem;
}

.button-container button:first-child {
  margin-right: 0.5rem;
}

.row {
  display: flex;
  flex-direction: row;
}

.small-container {
  padding-top: 1rem;
}

@media screen and (max-width: 667px) {
  .balance-header {
    display: flex;
    flex-direction: column;
  }

  .values-container {
    flex-direction: column;
    width: 100%;
  }

  .values-container .total-atoms__value {
    font-size: 28px;
    font-weight: 500;
    line-height: 32px;
  }

  .available-atoms,
  .rewards {
    padding: 0;
  }

  .total-atoms {
    padding: 1rem 0;
    text-align: center;
  }

  .button-container {
    width: 100%;
    padding: 1rem;
    border-top: 1px solid var(--bc);
  }

  .button-container button {
    width: 50%;
  }

  .small-container {
    display: flex;
    justify-content: space-evenly;
    padding: 1rem 0;
    text-align: center;
  }
}
</style>
