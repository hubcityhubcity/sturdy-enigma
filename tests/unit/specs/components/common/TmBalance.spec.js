import { shallowMount } from "@vue/test-utils"
import TmBalance from "common/TmBalance"

describe(`TmBalance`, () => {
  let wrapper, $store

  beforeEach(async () => {
    $store = {
      getters: {
        address: "cosmos1address",
        network: "test-network"
      }
    }

    wrapper = shallowMount(TmBalance, {
      mocks: {
        $store
      }
    })
    wrapper.setData({
      stakingDenom: "ATOM",
      overview: {
        totalStake: 3210,
        liquidStake: 1230,
        totalRewards: 1000.45
      }
    })
  })

  it(`show the balance header when signed in`, () => {
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.text()).toMatch(/Total ATOM.*3,210/s)
    expect(wrapper.text()).toMatch(/Available ATOM.*1,230/s)
    expect(wrapper.text()).toMatch(/Total Rewards.*1,000.45/s)
    expect(wrapper.text()).toContain("Total ATOM")
    expect(wrapper.text()).toContain("Total ATOM")
    expect(wrapper.text()).toContain("Total ATOM")
  })

  it(`show empty stakes when not signed in`, () => {
    // This *should* set overview to an empty Object, but the call is ignored?
    // Setting it to false has the desired effect.
    wrapper.setData({ overview: false })
    expect(wrapper.element).toMatchSnapshot()
    expect(wrapper.text()).toMatch(/Total ATOM.*--/s)
    expect(wrapper.text()).toMatch(/Available ATOM.*--/s)
    expect(wrapper.text()).not.toMatch(/Total Rewards/)
  })

  it(`opens send modal`, () => {
    const $refs = { SendModal: { open: jest.fn() } }
    wrapper.vm.$refs = $refs
    wrapper.find(".send-button").trigger("click")
    expect($refs.SendModal.open).toHaveBeenCalled()
  })

  it(`opens claim rewards modal`, () => {
    const $refs = { ModalWithdrawRewards: { open: jest.fn() } }
    wrapper.vm.$refs = $refs
    wrapper.find("#withdraw-btn").trigger("click")
    expect($refs.ModalWithdrawRewards.open).toHaveBeenCalled()
  })

  it(`disables claim rewards button when no rewards`, () => {
    wrapper.setData({
      overview: {
        totalRewards: 0
      }
    })
    const $refs = { ModalWithdrawRewards: { open: jest.fn() } }
    wrapper.vm.$refs = $refs
    wrapper.find("#withdraw-btn").trigger("click")
    expect($refs.ModalWithdrawRewards.open).not.toHaveBeenCalled()
  })
})
