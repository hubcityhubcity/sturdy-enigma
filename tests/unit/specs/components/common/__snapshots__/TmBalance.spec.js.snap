// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmBalance do not show available atoms when the user has none in the first place 1`] = `
<div
  class="balance-header"
>
  <div>
    <div
      class="header-container"
    >
      <h1>
        Your Portfolio
      </h1>
       
      <div
        class="buttons"
      >
        <tmbtn-stub
          class="send-button"
          type="secondary"
          value="Send"
        />
         
        <tmbtn-stub
          class="withdraw-rewards"
          id="withdraw-btn"
          value="Claim Rewards"
        />
         
        <button
          class="tutorial-button"
        >
          <span>
            Need some tokens?
          </span>
        </button>
         
        <div
          class="currency-selector"
        >
          <img
            alt="USD currency"
            class="currency-flag"
            src="/img/icons/currencies/usd.png"
          />
           
          <select>
            <option
              hidden="hidden"
              value=""
            >
              USD
            </option>
             
            <option
              value="USD"
            >
              USD
            </option>
             
            <option
              value="EUR"
            >
              EUR
            </option>
             
            <option
              value="GBP"
            >
              GBP
            </option>
             
            <option
              value="JPY"
            >
              JPY
            </option>
             
            <option
              value="CHF"
            >
              CHF
            </option>
             
            <option
              value="DKK"
            >
              DKK
            </option>
             
            <option
              value="NOK"
            >
              NOK
            </option>
             
            <option
              value="SEK"
            >
              SEK
            </option>
          </select>
        </div>
      </div>
    </div>
     
    <div
      class="table four-columns"
    >
      <div
        class="table-cell big title"
      >
        Token
      </div>
       
      <div
        class="table-cell title"
      >
        Rewards
      </div>
       
      <div
        class="table-cell title available"
      >
        Available
      </div>
       
      <div
        class="table-cell title actions"
      />
       
      <div
        class="table-cell big"
      >
        <img
          alt="ATOM currency"
          class="currency-flag"
          src="/img/icons/currencies/tom.png"
        />
         
        <div
          class="total-and-fiat"
        >
          <span
            class="total"
          >
            
            0
            ATOM
          
          </span>
           
          <!---->
        </div>
      </div>
       
      <div
        class="table-cell rewards"
      >
        <h2>
          
          +5
          ATOM
        
        </h2>
      </div>
       
      <div
        class="table-cell available"
      >
        <span
          class="available-amount"
        >
          
          0
          ATOM
        
        </span>
      </div>
       
      <div
        class="table-cell actions"
      >
        <div
          class="icon-button-container"
        >
          <button
            class="icon-button circle-send-button"
          >
            <i
              class="material-icons"
            >
              send
            </i>
          </button>
          <span>
            Send
          </span>
        </div>
      </div>
       
      <div
        class="table-cell big"
      >
        <img
          alt="TOKEN1 currency"
          class="currency-flag"
          src="/img/icons/currencies/oken1.png"
        />
         
        <div
          class="total-and-fiat"
        >
          <span
            class="total"
          >
            
              0
              TOKEN1
            
          </span>
           
          <!---->
        </div>
      </div>
       
      <div
        class="table-cell rewards"
      >
        <h2>
          
            +4.5
            TOKEN1
          
        </h2>
      </div>
       
      <div
        class="table-cell available"
      />
       
      <div
        class="table-cell actions"
      >
        <div
          class="icon-button-container"
        >
          <button
            class="icon-button"
          >
            <i
              class="material-icons"
            >
              send
            </i>
          </button>
          <span>
            Send
          </span>
        </div>
      </div>
    </div>
     
    <sendmodal-stub
      denoms="ATOM"
    />
     
    <modalwithdrawrewards-stub />
     
    <!---->
  </div>
</div>
`;

exports[`TmBalance show the balance header when signed in 1`] = `
<div
  class="balance-header"
>
  <div>
    <div
      class="header-container"
    >
      <h1>
        Your Portfolio
      </h1>
       
      <div
        class="buttons"
      >
        <tmbtn-stub
          class="send-button"
          type="secondary"
          value="Send"
        />
         
        <tmbtn-stub
          class="withdraw-rewards"
          id="withdraw-btn"
          value="Claim Rewards"
        />
         
        <button
          class="tutorial-button"
        >
          <span>
            Need some tokens?
          </span>
        </button>
         
        <div
          class="currency-selector"
        >
          <img
            alt="USD currency"
            class="currency-flag"
            src="/img/icons/currencies/usd.png"
          />
           
          <select>
            <option
              hidden="hidden"
              value=""
            >
              USD
            </option>
             
            <option
              value="USD"
            >
              USD
            </option>
             
            <option
              value="EUR"
            >
              EUR
            </option>
             
            <option
              value="GBP"
            >
              GBP
            </option>
             
            <option
              value="JPY"
            >
              JPY
            </option>
             
            <option
              value="CHF"
            >
              CHF
            </option>
             
            <option
              value="DKK"
            >
              DKK
            </option>
             
            <option
              value="NOK"
            >
              NOK
            </option>
             
            <option
              value="SEK"
            >
              SEK
            </option>
          </select>
        </div>
      </div>
    </div>
     
    <div
      class="table four-columns"
    >
      <div
        class="table-cell big title"
      >
        Token
      </div>
       
      <div
        class="table-cell title"
      >
        Rewards
      </div>
       
      <div
        class="table-cell title available"
      >
        Available
      </div>
       
      <div
        class="table-cell title actions"
      />
       
      <div
        class="table-cell big"
      >
        <img
          alt="ATOM currency"
          class="currency-flag"
          src="/img/icons/currencies/tom.png"
        />
         
        <div
          class="total-and-fiat"
        >
          <span
            class="total"
          >
            
            3,210
            ATOM
          
          </span>
           
          <!---->
        </div>
      </div>
       
      <div
        class="table-cell rewards"
      >
        <h2>
          
          +5
          ATOM
        
        </h2>
      </div>
       
      <div
        class="table-cell available"
      >
        <span
          class="available-amount"
        >
          
          1,230
          ATOM
        
        </span>
      </div>
       
      <div
        class="table-cell actions"
      >
        <div
          class="icon-button-container"
        >
          <button
            class="icon-button circle-send-button"
          >
            <i
              class="material-icons"
            >
              send
            </i>
          </button>
          <span>
            Send
          </span>
        </div>
      </div>
       
      <div
        class="table-cell big"
      >
        <img
          alt="TOKEN1 currency"
          class="currency-flag"
          src="/img/icons/currencies/oken1.png"
        />
         
        <div
          class="total-and-fiat"
        >
          <span
            class="total"
          >
            
              0
              TOKEN1
            
          </span>
           
          <!---->
        </div>
      </div>
       
      <div
        class="table-cell rewards"
      >
        <h2>
          
            +4.5
            TOKEN1
          
        </h2>
      </div>
       
      <div
        class="table-cell available"
      />
       
      <div
        class="table-cell actions"
      >
        <div
          class="icon-button-container"
        >
          <button
            class="icon-button"
          >
            <i
              class="material-icons"
            >
              send
            </i>
          </button>
          <span>
            Send
          </span>
        </div>
      </div>
    </div>
     
    <sendmodal-stub
      denoms="ATOM"
    />
     
    <modalwithdrawrewards-stub />
     
    <!---->
  </div>
</div>
`;
